### PR TITLE
Add TemporalFusionTransformer implementation in PyTorch

### PR DESCRIPTION
**Pull request by shchur**
*Description of changes:*

Unlike the MXNet implementation of TFT, the PyTorch versions of model/estimator are compatible with the dataset schema used by DeepAR. For example, if we construct the estimator with as follows
```python
estimator = TemporalFusionTransformerEstimator(
   ...,
   dynamic_dims=[2, 5, 7],
   dynamic_cardinalities=[6, 5, 5, 4, 2],
   past_dynamic_dims=[3, 3],
   past_dynamic_cardinalities=[5, 2, 4],
)
```
it will expect to receive a dataset, where each time series has keys
- `"feat_dynamic_real"`: shape `[..., sum(dynamic_dims)]` (inside the network this will feature get partitioned into 3 chunks of dim 2, 5, 7)
- `"feat_dynamic_cat"`: shape `[..., len(dynamic_cardinalities)]`
- `"past_feat_dynamic_real"`: shape `[..., sum(past_dynamic_dims)]`
- `"past_feat_dynamic_cat"`: shape `[..., len(past_dynamic_cardinalities)]`

*To do:*
- [x] Move `QuantileOutput` and `TFTInstanceSplitter` to a another folder?
- [x] Benchmarking
- [x] Add tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by kashif** [comment]
very cool! I too have a version of TFT which is auto-regressive and the output can be a prob distrb class, rather than quantiles.... kindly have a look here: https://github.com/kashif/pytorch-transformer-ts/tree/main/tft

**Comment by shchur** [comment]
Thank you, @kashif! Some parts of this PR are based on your code (there is an acknowledgement in the docstring), thanks a lot for making it available! My main focus here was to add support for all types of dynamic features, plus we want to stick to quantile output, since faster inference is critical for some downstream applications in AutoGluon-TS.

File: src/gluonts/model/forecast_generator.py [hunk 1]
```diff
@@ -114,8 +114,9 @@ def __call__(
         **kwargs
     ) -> Iterator[Forecast]:
         for batch in inference_data_loader:
-            inputs = [batch[k] for k in input_names]
-            outputs = predict_to_numpy(prediction_net, inputs)
+            # TODO: Change inputs to dict to avoid assuming matching order in input_names / model signature?
+            inputs = [batch.get(k) for k in input_names]
+            outputs = prediction_net(*inputs).cpu().detach().numpy()
```
Comment (lostella) [inline src/gluonts/model/forecast_generator.py]: I think this works only with some type of tensors, right? `predict_to_numpy` should take care of this (although it's missing `.detach()`?)

File: src/gluonts/torch/model/tft/transformation.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from gluonts.mx.model.tft._transform import TFTInstanceSplitter
```
Comment (lostella) [inline src/gluonts/torch/model/tft/transformation.py]: As you point out in the PR, it would probably be better to move this somewhere else, or otherwise just copy the code here. Otherwise this may put a dependency on `mxnet`.

Comment (lostella) [reply src/gluonts/torch/model/tft/transformation.py]: I'm also wondering which the model needs such a dedicated component? We could probably simplify how this step works across the board. But maybe let's revisit this separately and just have the code copied here for now?

File: src/gluonts/torch/modules/scaler.py [hunk 3]
```diff
@@ -111,3 +111,52 @@ def forward(
...
+        self.keepdim = keepdim
+        self.register_buffer("minimum_scale", torch.tensor(minimum_scale))
+
+    def forward(
+        self, data: torch.Tensor, weights: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        assert (
+            data.shape == weights.shape
+        ), "data and weights must have same shape"
+        with torch.no_grad():
```
Comment (lostella) [inline src/gluonts/torch/modules/scaler.py:163]: Also to be addressed separately, but: all these scalers don't really have learnable parameters, so they could really be functions or very simple `__call__`able dataclasses. Thoughts?

Comment (kashif) [reply src/gluonts/torch/modules/scaler.py:163]: I agree

File: src/gluonts/model/forecast_generator.py (hunk 1 above)
Comment (shchur) [reply src/gluonts/model/forecast_generator.py]: My bad, this change is actually unnecessary. 

There is still the question of whether `inputs` should be a `dict` instead of list to remove the implicit assumption that the of arguments in `input_names` matches the order expected by `forward` method of the model.

File: src/gluonts/torch/model/tft/transformation.py (hunk 2 above)
Comment (shchur) [reply src/gluonts/torch/model/tft/transformation.py]: I've moved `TFTInstanceSplitter` to `gluonts.transform.split` & added a docstring.

File: src/gluonts/torch/modules/scaler.py (hunk 3 above)
Comment (shchur) [reply src/gluonts/torch/modules/scaler.py:163]: I agree too. We could also modify `MeanScaler` and `NOPScaler` to also return `data, loc, scale` for consistency (currently they only return `data, scale`).

Comment (kashif) [reply src/gluonts/torch/modules/scaler.py:163]: I can do that and change it for every model... and also add log1p of loc as a feature together with log scale?

**Comment by timoschowski** [comment]
any chance that this is going in before the Sphere course starts on Jan 17? That would be rather awesome!

**Comment by shchur** [comment]
Here are some benchmarking results for PyTorch vs. MXNet implementations of TFT on randomly chosen 12 datasets. Winrate is computed as % of metrics where the PyTorch version is better (assuming higher-is-better for coverage/related metrics, and lower-is-better for the rest). 

The PyTorch version seems to consistently be 2-4x faster, but the performance is rather different between the two models - with PyTorch being better in 8/12 cases. Is this an acceptable difference @lostella?

<details>

- max_epochs = 50
- batch_size = 64
- context_length = 64
- quantile_levels = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
- no validation set - just train on the train set until max_epochs is reached
- both models trained on 1xV100 GPU

</details>

| dataset                       |   train_time_torch |  train_time_mxnet |   torch_winrate |
|:------------------------------|--------:|--------:|----------------:|
| australian_electricity_demand |     555 |    1007 |            70   |
| bitcoin                       |     178 |     598 |            77.5 |
| fred_md                       |     150 |     488 |            85   |
| kdd_2018                      |     206 |     694 |            30   |
| m3_monthly                    |     149 |     494 |            10   |
| m4_hourly                     |     194 |     627 |            95   |
| m4_weekly                     |     172 |     483 |            27.5 |
| restaurant                    |     177 |     611 |            77.5 |
| rideshare                     |     205 |     672 |            57.5 |
| rossmann                      |     177 |     642 |            22.5 |
| tourism_monthly               |     153 |     543 |            67.5 |
| walmart                       |     169 |     595 |            57.5 |

Raw metrics for each dataset - [benchmark_tft.zip](https://github.com/awslabs/gluonts/files/10411229/benchmark_tft.zip)

File: src/gluonts/torch/model/tft/estimator.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    num_heads
+        Number of attention heads in self-attention layer in the decoder.
+    hidden_dim
+        Size of the LSTM & transformer hidden states.
+    variable_dim
+        Size of the feature embeddings.
+    static_dims
+        Sizes of the real-valued static features.
+    dynamic_dims
+        Sizes of the real-valued dynamic features that are known in the future.
```
Comment (lostella) [inline src/gluonts/torch/model/tft/estimator.py:107]: I’m a bit confused: why does one need to specify multiple sizes?

**Review by lostella — COMMENTED** [review]
I think it would be nice to also add this to this smoke test: https://github.com/awslabs/gluonts/blob/9e5c4e90693085a4067ba77e43128cbbc4eccbf5/test/torch/model/test_modules.py#L88

File: src/gluonts/torch/model/tft/module.py [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+
+    Partially based on the implementation in github.com/kashif/pytorch-transformer-ts.
+    """
+
+    @validated()
+    def __init__(
+        self,
+        context_length: int,
+        prediction_length: int,
+        d_past_feat_dynamic_real: List[int],
```
Comment (lostella) [inline src/gluonts/torch/model/tft/module.py]: I’m wondering whether it would make sense to have defaults for most of these. For example, in case one has no additional features (only target series) what would these be set to? I’m assuming some minimal values and then constant dummy tensors would be given to `forward` in place of features?

Comment (shchur) [reply src/gluonts/torch/model/tft/module.py]: Added defaults here + modified the docstring

File: src/gluonts/torch/model/tft/estimator.py (hunk 4 above)
Comment (shchur) [reply src/gluonts/torch/model/tft/estimator.py:107]: I've added an explanation in the docstring for `TemporalFusionTransformerEstimator`. Does this make sense?

```
    TFT internally performs feature selection when making forecasts. For this
    reason, the dimensions of real-valued features can be grouped together if
    they correspond to the same variable (e.g., treat weather features as a
    one feature and holiday indicators as another feature).

    For example, if the dataset contains key "feat_static_real" with shape
    [batch_size, 3], we can, e.g.,
    - set ``static_dims = [3]`` to treat all five dimensions as a single feature
    - set ``static_dims = [1, 1, 1]`` to treat each dimension as a separate feature
    - set ``static_dims = [2, 1]`` to treat the first two dims as a single feature

    See ``gluonts.torch.model.tft.TemporalFusionTransformerModel.input_shapes``
    for more details on how the model configuration corresponds to the expected
    input shapes.
```

This way the TFT estimator / model is compatible with the DeepAR schema if we set
```python
d_feat_dynamic_real = [num_feat_dynamic_real],
```

File: src/gluonts/torch/model/tft/module.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    def __init__(
+        self,
+        context_length: int,
+        prediction_length: int,
+        d_feat_static_real: List[int] = None,  #  Defaults to [1]
+        c_feat_static_cat: List[int] = None,  #  Defaults to [1]
+        d_feat_dynamic_real: List[int] = None,  # Defaults to [1]
+        c_feat_dynamic_cat: List[int] = None,  # Defaults to []
+        d_past_feat_dynamic_real: List[int] = None,  # Defaults to []
+        c_past_feat_dynamic_cat: List[int] = None,  # Defaults to []
```
Comment (lostella) [inline src/gluonts/torch/model/tft/module.py]: These should be `Optional[List[int]]`

File: src/gluonts/torch/model/tft/estimator.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+    Estimator class to train a Temporal Fusion Transformer (TFT) model, as described in [LAL+21]_.
+
+    TFT internally performs feature selection when making forecasts. For this
+    reason, the dimensions of real-valued features can be grouped together if
+    they correspond to the same variable (e.g., treat weather features as a
+    one feature and holiday indicators as another feature).
+
+    For example, if the dataset contains key "feat_static_real" with shape
+    [batch_size, 3], we can, e.g.,
+    - set ``static_dims = [3]`` to treat all five dimensions as a single feature
```
Comment (lostella) [inline src/gluonts/torch/model/tft/estimator.py]:
```suggestion
    - set ``static_dims = [3]`` to treat all three dimensions as a single feature
```

**Review by lostella — CHANGES_REQUESTED** [review]

**Comment by shchur** [comment]
Resolved 👍

File: src/gluonts/torch/modules/scaler.py (hunk 3 above)
Comment (lostella) [reply src/gluonts/torch/modules/scaler.py:163]: I guess we can make incremental changes: first make these not-`torch.nn.Module` objects, then possibly change their interface. The `loc` change will be breaking, so let's discuss that separately maybe.

**Comment by lostella** [comment]
@shchur did you get those results with default settings? I'm interested in `context_length` in particular, which appears to default to `prediction_length` when not specified (in the estimator).

**Comment by shchur** [comment]
These were obtained using `context_length = 64` (other hyperparameters also mentioned in the dropdown "Details" in the original message). The [reference implementation](https://github.com/google-research/google-research/blob/master/tft/data_formatters/base.py) uses a different `context_length` for each dataset, so I just set it to be the same as in the MXNet implementation of TFT in GluonTS.

Though, most likely, the choice of `context_length = prediction_length` is very poor, do you think we should change it to something else?

**Comment by lostella** [comment]
> Though, most likely, the choice of context_length = prediction_length is very poor, do you think we should change it to something else?

Exactly, it should probably be much longer. Maybe let's do that separately from this PR, we could think of either a larger multiple of `prediction_length` or some heuristics based on the frequency of the data, or a combination of both.

**Review by lostella — APPROVED** [review]

**Comment by kashif** [comment]
awesome!
